src/game/stage/item_box.c

```diff
@@ -0,0 +1,518 @@
+#include "global.h"
+#include "core.h"
+#include "malloc_vram.h"
+#include "sprite.h"
+#include "task.h"
+#include "trig.h"
+#include "lib/m4a.h"
+#include "game/game.h"
+#include "game/entity.h"
+
+#include "constants/animations.h"
+#include "constants/songs.h"
+#include "constants/zones.h"
+
+typedef struct {
+    /* 0x00 */ SpriteBase base;
+    /* 0x0C */ Sprite s;
+    /* 0x3C */ Sprite item;
+    /* 0x6C */ s32 x;
+    /* 0x70 */ s32 y;
+    /* 0x74 */ s16 unk74;
+    /* 0x76 */ u8 kind;
+    /* 0x77 */ u8 frames;
+    /* 0x78 */ bool8 unk78;
+} Entity_ItemBox; /* size: 0x7C */
+
+void sub_800B1AC(Entity_ItemBox *);
+void sub_800B580(Entity_ItemBox *, bool32);
+void sub_800B860(Entity_ItemBox *, bool32);
+void Task_800B704(void);
+void Task_800B780(void);
+void Task_800B7D0(void);
+void TaskDestructor_800B80C(struct Task *);
+void sub_800B828(Entity_ItemBox *);
+bool32 sub_800B8AC(Entity_ItemBox *);
+bool32 sub_800B8F4(Entity_ItemBox *);
+void Task_800B950(void);
+void sub_800B9A0(Entity_ItemBox *);
+
+const u16 ItemBox_MysteryIcons[13][3] = {
+    { SA2_ANIM_ITEMBOX_TYPE, 0, 4 },  { SA2_ANIM_ITEMBOX_TYPE, 5, 4 },
+    { SA2_ANIM_ITEMBOX_TYPE, 6, 4 },  { SA2_ANIM_ITEMBOX_TYPE, 7, 4 },
+    { SA2_ANIM_ITEMBOX_TYPE, 8, 4 },  { SA2_ANIM_ITEMBOX_TYPE, 9, 4 },
+    { SA2_ANIM_ITEMBOX_TYPE, 10, 4 }, { SA2_ANIM_ITEMBOX_TYPE, 11, 4 },
+    { SA2_ANIM_ITEMBOX_TYPE, 13, 4 }, { SA2_ANIM_ITEMBOX_TYPE, 14, 4 },
+    { SA2_ANIM_ITEMBOX_TYPE, 15, 4 }, { SA2_ANIM_ITEMBOX_TYPE, 16, 4 },
+    { SA2_ANIM_ITEMBOX_TYPE, 12, 4 },
+};
+
+const u16 ItemBox_1UpIcons[5][3] = {
+    { SA2_ANIM_ITEMBOX_TYPE, CHARACTER_SONIC, 4 },
+    { SA2_ANIM_ITEMBOX_TYPE, CHARACTER_CREAM, 4 },
+    { SA2_ANIM_ITEMBOX_TYPE, CHARACTER_TAILS, 4 },
+    { SA2_ANIM_ITEMBOX_TYPE, CHARACTER_KNUCKLES, 4 },
+    { SA2_ANIM_ITEMBOX_TYPE, CHARACTER_AMY, 4 },
+};
+
+const u16 gUnknown_080D51E4[8] = { 5, 13, 4, 13, 3, 13, 6, 13 };
+
+const u16 gUnknown_080D51F4[] = { 8, 9, 10, 11 };
+const u8 gUnknown_080D51FC[] = { 1, 5, 10, 30, 50, 0, 0, 0 };
+
+void CreateEntity_ItemBox(MapEntity *me, u16 spriteRegionX, u16 spriteRegionY,
+                          u8 spriteY)
+{
+    struct Task *t;
+    Entity_ItemBox *itembox;
+    Sprite *s;
+    if (gGameMode == GAME_MODE_TIME_ATTACK || gGameMode == GAME_MODE_BOSS_TIME_ATTACK) {
+        if (me->index == 0) {
+            goto CreateEntity_Itembox_defer;
```

Comment:
Interesting. Did you try this with a return?

Reply:
Yeah I did, but initially it didn't work.

But I tried again and it does, thanks!

Reply:
Though I feel like that the goto is "cleaner" than setting the ME init at the top and the bottom, idk.

Reply:
Nah, I think it's best to write what the developers probably wrote. So far this seems to be a pattern